core/field.js

```diff
@@ -633,39 +638,114 @@ Blockly.Field.prototype.forceRerender = function() {
 };
 
 /**
- * By default there is no difference between the human-readable text and
- * the language-neutral values.  Subclasses (such as dropdown) may define this.
- * @return {string} Current value.
+ * Get the current value of the field.
+ * @return {*} Current value.
  */
 Blockly.Field.prototype.getValue = function() {
-  return this.getText();
+  return this.value_;
 };
 
 /**
- * By default there is no difference between the human-readable text and
- * the language-neutral values.  Subclasses (such as dropdown) may define this.
- * @param {string} newValue New value.
+ * Used to change the value of the field. Handles validation and events.
+ * Subclasses should override doClassValidation_ and doValueUpdate_ rather
+ * than this method.
+ * @param {*} newValue New value.
  */
 Blockly.Field.prototype.setValue = function(newValue) {
+  var doLogging = true;
   if (newValue === null) {
-    // No change if null.
+    doLogging && console.log('null, return');
+    // Not a valid value to check.
     return;
   }
-  // Validate input.
-  var validated = this.callValidator(newValue);
-  if (validated !== null) {
-    newValue = validated;
+
+  newValue = this.doClassValidation_(newValue);
+  if (newValue === null) {
+    doLogging && console.log('invalid, return');
+    this.doValueInvalid_();
+    if (this.isDirty_) {
+      this.forceRerender();
+    }
+    return;
+  }
+
+  var userValidator = this.getValidator();
+  if (userValidator) {
+    newValue = userValidator.call(this, newValue);
+    if (newValue === null) {
+      doLogging && console.log('invalid, return');
+      this.doValueInvalid_();
+      if (this.isDirty_) {
+        this.forceRerender();
+      }
+      return;
+    }
   }
-  // Check for change.
   var oldValue = this.getValue();
   if (oldValue == newValue) {
+    doLogging && console.log('same, return');
+    // No change.
     return;
   }
+
   if (this.sourceBlock_ && Blockly.Events.isEnabled()) {
     Blockly.Events.fire(new Blockly.Events.BlockChange(
         this.sourceBlock_, 'field', this.name, oldValue, newValue));
   }
-  this.setText(newValue);
+  this.doValueUpdate_(newValue);
+  if (this.isDirty_) {
+    // Call setText instead of calling forceRerender
+    // directly for backwards compatibility.
+    this.setText(newValue);
+    if (this.isDirty_) {
+      // setText must not have called forceRerender, either because it was
+      // overridden, or text_ was set inside doValueUpdate_. Try again.
+      this.forceRerender();
+    }
+  }
+  doLogging && console.log(this.value_);
+};
+
+/**
+ * A generic value possessed by the field.
+ * Should generally be non-null, only null when the field is created.
+ * @type {*}
+ * @protected
+ */
+Blockly.Field.prototype.value_ = null;
+
+/**
+ * Used to validate a value. Returns input by default. Can be overridden by
+ * subclasses, see FieldDropdown.
+ * @param {*} newValue The value to be validated.
+ * @return {*} The validated value, same as input by default.
```

Comment:
As I recall, the possible return values are null, undefined, or the validated value, where null means it's invalid and undefined means...something.  Is that accurate?

Either way, please update the comment to say what null means.

Reply:
Null means its invalid, which by default means the field reverts to its previous value.

Undefined doesn't mean anything at the moment, but most fields will revert to a default value (default meaning the value they would have if they had been constructed with no arguements, not any value they /were/ constructed with) when passed a value they can't parse (such as undefined, NaN, 'bad', etc).

This seems like a bug, since not all fields work this way. What I want to go back through and do is have fields' class validators validate anything they can't parse (undefined, NaN, 'bad', etc) to null (invalid). Any default values would be handled in the constructor.

But that's a bit besides the point, if we want to add more special cases besides 'invalid' I would prefer to move to using constants rather than keywords, so it's more extendable in the future.